File: lib/src/uploader.ts
import type { S3Client } from "@aws-sdk/client-s3";
import { Upload } from "@aws-sdk/lib-storage";
import { createReadStream } from "node:fs";
import { readdir } from "node:fs/promises";
import path from "node:path";

export interface UploadDirectoryOptions {
  client: S3Client;
  bucket: string;
  /** Object key prefix in the destination bucket. Should end with "/". */
  keyPrefix: string;
  localDir: string;
  /** Concurrent uploads. Default: 8. */
  concurrency?: number;
}

/**
 * Uploads every file under `localDir` to `bucket` under `keyPrefix`,
 * preserving relative paths. Sets HLS-aware Content-Type headers.
 * Uses lib-storage Upload so large segments use multipart automatically.
 */
export async function uploadDirectory(opts: UploadDirectoryOptions): Promise<string[]> {
  const { client, bucket, keyPrefix, localDir } = opts;
  const concurrency = opts.concurrency ?? 8;

  const files: string[] = [];
  for await (const file of walkDir(localDir)) {
    files.push(file);
  }

  const keys: string[] = new Array(files.length);
  await pmap(files, concurrency, async (filePath, i) => {
    const rel = path.relative(localDir, filePath);
    const key = `${keyPrefix}${rel.split(path.sep).join("/")}`;

    const upload = new Upload({
      client,
      params: {
        Bucket: bucket,
        Key: key,
        Body: createReadStream(filePath),
        ContentType: contentTypeFor(rel),
      },
    });
    await upload.done();
    keys[i] = key;
  });

  return keys;
}

async function* walkDir(dir: string): AsyncIterable<string> {
  for (const entry of await readdir(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      yield* walkDir(full);
    } else if (entry.isFile()) {
      yield full;
    }
  }
}

export function contentTypeFor(filename: string): string {
  if (filename.endsWith(".m3u8")) return "application/vnd.apple.mpegurl";
  if (filename.endsWith(".m4s")) return "video/iso.segment";
  if (filename.endsWith(".mp4")) return "video/mp4";
  if (filename.endsWith(".ts")) return "video/mp2t";
  if (filename.endsWith(".vtt")) return "text/vtt";
  if (filename.endsWith(".json")) return "application/json";
  return "application/octet-stream";
}

async function pmap<T, R>(
  items: readonly T[],
  concurrency: number,
  fn: (item: T, index: number) => Promise<R>,
): Promise<R[]> {
  const results: R[] = new Array(items.length);
  let next = 0;
  async function worker(): Promise<void> {
    for (;;) {
      const i = next++;
      if (i >= items.length) return;
      results[i] = await fn(items[i]!, i);
    }
  }
  const workers = Array.from({ length: Math.min(concurrency, items.length) }, () => worker());
  await Promise.all(workers);
  return results;
}

File: core/uploader.go
package core

import (
	"context"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

// ContentTypeFor returns an HLS-aware content type for a filename.
func ContentTypeFor(filename string) string {
	switch {
	case strings.HasSuffix(filename, ".m3u8"):
		return "application/vnd.apple.mpegurl"
	case strings.HasSuffix(filename, ".m4s"):
		return "video/iso.segment"
	case strings.HasSuffix(filename, ".mp4"):
		return "video/mp4"
	case strings.HasSuffix(filename, ".ts"):
		return "video/mp2t"
	case strings.HasSuffix(filename, ".vtt"):
		return "text/vtt"
	case strings.HasSuffix(filename, ".json"):
		return "application/json"
	default:
		return "application/octet-stream"
	}
}

// UploadDirectory uploads every file under localDir to bucket under keyPrefix,
// preserving relative paths, with HLS-aware content types. Large segments use
// multipart automatically via the upload manager.
func UploadDirectory(ctx context.Context, client *s3.Client, bucket, keyPrefix, localDir string, concurrency int) ([]string, error) {
	if concurrency <= 0 {
		concurrency = 8
	}
	var files []string
	if err := filepath.WalkDir(localDir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if !d.IsDir() {
			files = append(files, path)
		}
		return nil
	}); err != nil {
		return nil, err
	}

	uploader := manager.NewUploader(client)
	keys := make([]string, len(files))

	ctx, cancel := context.WithCancel(ctx)
	defer cancel()
	var (
		wg       sync.WaitGroup
		mu       sync.Mutex
		firstErr error
		sem      = make(chan struct{}, concurrency)
	)
	fail := func(err error) {
		mu.Lock()
		if firstErr == nil {
			firstErr = err
			cancel()
		}
		mu.Unlock()
	}

	for i, path := range files {
		sem <- struct{}{}
		wg.Add(1)
		go func(i int, path string) {
			defer wg.Done()
			defer func() { <-sem }()
			rel, err := filepath.Rel(localDir, path)
			if err != nil {
				fail(err)
				return
			}
			key := keyPrefix + filepath.ToSlash(rel)
			f, err := os.Open(path)
			if err != nil {
				fail(err)
				return
			}
			defer f.Close()
			if _, err := uploader.Upload(ctx, &s3.PutObjectInput{
				Bucket:      aws.String(bucket),
				Key:         aws.String(key),
				Body:        f,
				ContentType: aws.String(ContentTypeFor(rel)),
			}); err != nil {
				fail(fmt.Errorf("upload %s: %w", key, err))
				return
			}
			keys[i] = key
		}(i, path)
	}
	wg.Wait()
	if firstErr != nil {
		return nil, firstErr
	}
	return keys, nil
}
